build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1288,6 +1670,228 @@ message WaitExecutionRequest {
+  int32 page_size = 4;
+
+  // A page token, which must be a value received in a previous
+  // [GetPotentialActionResultsResponse][build.bazel.remote.execution.v2.GetPotentialActionResultsResponse].
+  // If present, the server will use that token as an offset, returning only
+  // that page and the ones that succeed it.
+  string page_token = 5;
+}
+
+// Returns a set of zero or more
+// (pathset, observed-input digest, `ActionResult` digest) results
+// from a two-phase cache lookup call to
+// [ActionCache.GetPotentialActionResults][build.bazel.remote.execution.v2.ActionCache.GetPotentialActionResults].
+// These results are compared to the local filesystem and, on a match to the
+// observed-input digest, the associated `ActionResult` is used to retrieve
+// cached outputs.
+message GetPotentialActionResultsResponse {
+  // A set of possible matches to the client filesystem along with a reference to build outputs.
+  repeated ObservedPathSetAndInputDigest potential_action_matches = 1;
+
+  // If present, signifies that there are more results which the client can
+  // retrieve by passing this as the page_token in a subsequent
+  // [request][build.bazel.remote.execution.v2.GetPotentialActionResults].
+  // If empty, signifies that this is the last page of results.
+  string next_page_token = 2;
+}
+
+// The ordered file reads and probes and directory enumerations used in
+// two-phase caching to compare to the client filesystem when seeking a match
+// from an action fingerprint formed from potentially incomplete
+// (underspecified) inputs to a set of outputs.
+//
+// When comparing to the filesystem, a corresponding observed-input digest is
+// initialized. Each pathset entry is examined against the filesystem, and
+// depending on the outcome of the comparison, new hash information is added
+// to the observed-input digest. At the end of pathset enumeration the
+// observed-input digest is compared to the observed-input digest provided in
+// the two-phase cache lookup (`observed_input_digest` in
+// [ObservedPathSetAndInputDigest][build.bazel.remote.execution.v2.ObservedPathSetAndInputDigest])
+// and if they match, this is a cache hit, and the resulting
+// [ActionResult][build.bazel.remote.execution.v2.ActionResult] should be
+// retrieved and used.
+//
+// Generating a directory hash: Create a list of the names of directory members
+// (files and directories) that pass the directory enumeration or
+// `ObservedAccessFileNames` filters. Lexicographically sort the names by code
+// point or, equivalently, by the UTF-8 bytes. Join the sorted strings together
+// with the ',' character to form a comma-delimited list; note for a
+// multi-member list this means that there is no tail ',', e.g. for
+// a.h and b.h the list string is "a.h,b.h". Convert this string to uppercase.
+// Hash the UTF-8 string bytes.
+//
+// Discussion of how to generate pathsets from action execution filesystem
```

Comment:
It can be reasonable to link out to extra content, but this API must stand alone without it (e.g. when that link eventually goes stale). Please make sure a reader without access to this wiki page will have all necessary information to figure out the API here.

Reply:
Alexis had a comment about how the algo works (above), and also indicated this content is not understandable without a lot more context. Do you think adding in examples to the doc comments works?